test/adaptive_load/plugin_util_test.cc

```diff
@@ -0,0 +1,377 @@
+      : value_from_config_proto_{config.threshold()} {}
+  double EvaluateMetric(double) const override { return 1.0; }
+  const double value_from_config_proto_;
+};
+
+/**
+ * A factory that creates a TestScoringFunction from a LinearScoringFunctionConfig (see
+ * TestScoringFunction constructor).
+ */
+class TestScoringFunctionConfigFactory : public ScoringFunctionConfigFactory {
+public:
+  std::string name() const override { return "nighthawk.test-scoring-function"; }
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<nighthawk::adaptive_load::LinearScoringFunctionConfig>();
+  }
+  absl::Status ValidateConfig(const Envoy::Protobuf::Message& message) override {
+    return DoValidateConfig(message);
+  }
+  ScoringFunctionPtr createScoringFunction(const Envoy::Protobuf::Message& message) override {
+    const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+    nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+    Envoy::MessageUtil::unpackTo(any, config);
+    return std::make_unique<TestScoringFunction>(config);
+  }
+};
+
+REGISTER_FACTORY(TestScoringFunctionConfigFactory, ScoringFunctionConfigFactory);
+
+/**
+ * MetricsPlugin for testing.
+ */
+class TestMetricsPlugin : public MetricsPlugin {
+public:
+  // Any plugin in the adaptive load system can freely choose an arbitrary single proto as its
+  // config type. We use LinearScoringFunctionConfig for all plugins in this test.
+  TestMetricsPlugin(const nighthawk::adaptive_load::LinearScoringFunctionConfig& config)
+      : value_from_config_proto_{config.threshold()} {}
+  Envoy::StatusOr<double> GetMetricByName(absl::string_view) override { return 5.0; }
+  const std::vector<std::string> GetAllSupportedMetricNames() const override { return {}; }
+  const double value_from_config_proto_;
+};
+
+/**
+ * A factory that creates a TestMetricsPlugin from a LinearScoringFunctionConfig (see
+ * TestInputVariableSetter constructor).
+ */
+class TestMetricsPluginConfigFactory : public MetricsPluginConfigFactory {
+public:
+  std::string name() const override { return "nighthawk.test-metrics-plugin"; }
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<nighthawk::adaptive_load::LinearScoringFunctionConfig>();
+  }
+  absl::Status ValidateConfig(const Envoy::Protobuf::Message& message) override {
+    return DoValidateConfig(message);
+  }
+  MetricsPluginPtr createMetricsPlugin(const Envoy::Protobuf::Message& message) override {
+    const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+    nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+    Envoy::MessageUtil::unpackTo(any, config);
+    return std::make_unique<TestMetricsPlugin>(config);
+  }
+};
+
+REGISTER_FACTORY(TestMetricsPluginConfigFactory, MetricsPluginConfigFactory);
+
+/**
+ * StepController for testing.
+ */
+class TestStepController : public StepController {
```

Comment:
this seems pretty close to mocking a step controller. It works for me as is, but out of curiosity, I do wonder if there was an explicit choice made here to not use the mocking and instead pull it off like this?

Reply:
I couldn't find a way to cleanly test all the behavior using mocks (specifically, the plugin util must propagate its input args to the plugin constructor). But admittedly the fakes in this test are not very good. Filed #448 to create better fakes that can be used across the tests. I already have a much better fake StepController in one of the upcoming PRs, and I can turn it into a library and reuse it in this test. 

This PR is a prerequisite for another PR that will take some time to review, so can we merge this with these bad fakes for now, rather than changing to mocks which can't easily test the full behavior, then replace these fakes with the better fakes when they're ready?

Reply:
Sounds great to me, thanks for sharing context.